kernel/comps/framebuffer/src/console.rs

```diff
@@ -0,0 +1,189 @@
+#[derive(Debug)]
+pub struct FramebufferConsole {
+    state: SpinLock<ConsoleState, LocalIrqDisabled>,
+}
+
+pub static CONSOLE_NAME: &str = "Framebuffer-Console";
+
+pub static FRAMEBUFFER_CONSOLE: Once<Arc<FramebufferConsole>> = Once::new();
+
+pub(crate) fn init() {
+    let Some(fb) = FRAMEBUFFER.get() else {
+        log::warn!("Framebuffer not initialized");
+        return;
+    };
+
+    FRAMEBUFFER_CONSOLE.call_once(|| Arc::new(FramebufferConsole::new(fb.clone())));
+}
+
+impl AnyConsoleDevice for FramebufferConsole {
+    fn send(&self, buf: &[u8]) {
+        self.state.lock().send_buf(buf);
+    }
+
+    fn register_callback(&self, _: &'static ConsoleCallback) {
+        // Unsupported, do nothing.
+    }
+}
+
+impl FramebufferConsole {
+    /// Creates a new framebuffer console.
+    pub fn new(framebuffer: Arc<FrameBuffer>) -> Self {
+        let bytes = alloc::vec![0u8; framebuffer.size()];
+        Self {
+            state: SpinLock::new(ConsoleState {
+                x_pos: 0,
+                y_pos: 0,
+                fg_color: Pixel::WHITE,
+                bg_color: Pixel::BLACK,
+                bytes,
+                backend: framebuffer,
+            }),
+        }
+    }
+
+    /// Returns the current cursor position.
+    pub fn cursor(&self) -> (usize, usize) {
+        let state = self.state.lock();
+        (state.x_pos, state.y_pos)
+    }
+
+    /// Sets the cursor position.
+    pub fn set_cursor(&self, x: usize, y: usize) -> Result<()> {
+        let mut state = self.state.lock();
+        if x > state.backend.width() - FONT_WIDTH || y > state.backend.height() - FONT_HEIGHT {
+            log::warn!("Invalid framebuffer cursor position: ({}, {})", x, y);
+            return Err(Error::InvalidArgs);
+        }
+        state.x_pos = x;
+        state.y_pos = y;
+        Ok(())
+    }
+
+    /// Returns the foreground color.
+    pub fn fg_color(&self) -> Pixel {
+        self.state.lock().fg_color
+    }
+
+    /// Sets the foreground color.
+    pub fn set_fg_color(&self, val: Pixel) {
+        self.state.lock().fg_color = val;
+    }
+
+    /// Returns the background color.
+    pub fn bg_color(&self) -> Pixel {
+        self.state.lock().bg_color
+    }
+
+    /// Sets the background color.
+    pub fn set_bg_color(&self, val: Pixel) {
+        self.state.lock().bg_color = val;
+    }
+}
+
+#[derive(Debug)]
+struct ConsoleState {
+    x_pos: usize,
+    y_pos: usize,
+    fg_color: Pixel,
+    bg_color: Pixel,
+    bytes: Vec<u8>,
+    backend: Arc<FrameBuffer>,
+}
+
+impl ConsoleState {
+    fn carriage_return(&mut self) {
+        self.x_pos = 0;
+    }
+
+    fn newline(&mut self) {
+        if self.y_pos >= self.backend.height() - FONT_HEIGHT {
+            self.shift_lines_up();
+        }
+        self.y_pos += FONT_HEIGHT;
+        self.x_pos = 0;
+    }
+
+    fn shift_lines_up(&mut self) {
+        let offset = self.backend.width() * self.backend.pixel_format().nbytes() * FONT_HEIGHT;
```

Comment:
The new code is very clear as it clearly states that we need to multiply with `FONT_HEIGHT`.

Perhaps now `self.backend.calc_offset(0, FONT_HEIGHT).as_usize()` is even better?